TT-5: Make JAWS Work with GPII

xpheal opened the pull request:
@amb26 , Here's the new pull request. I've created a new branch for JAWS, will work on other solutions soon.

kaspermarkus commented on testData/solutions/jaws.json:
```diff
@@ -0,0 +1,123 @@
+{
```
Unless you think otherwise, I believe this is not used and can safely be deleted.

kaspermarkus commented on testData/solutions/win32.json:
```diff
@@ -9,16 +9,114 @@
+                    "options\\.UseWindowsKeyboardHook": "http://registry\\.gpii\\.net/common/useWindowsKeyboardHook",
+                    "options\\.SayCursorShapeChange": "http://registry\\.gpii\\.net/common/sayCursorShapeChange",
+                    "options\\.SayNumericDates": "http://registry\\.gpii\\.net/common/sayNumericDates",
+                    "options\\.SayDollars": "http://registry\\.gpii\\.net/common/sayDollars"
```
For all the above settings/preferences, it's worth considering which are useable (and sensible) beyond Jaws. No matter which mappings you put in the above, it will always be possible for users to have the above settings in their NP set (eg. if you didn't have the JAWSInsertKey map of line 68, the user would still be able to store the JAWSInsertKey setting in their preferences set).

Generally, you should go through the list above and do the following:
* Consider whether the setting is sensible to have as a general preference that can be used beyond Jaws.
** If it is not, delete it
** If it is keep it and add it to this spreadsheet: https://docs.google.com/spreadsheets/d/1Cj9YTzM6usgd96XFn60TGN6bfMYTLCgWVMycoA4JhA8/edit#gid=0 along with the following information:
*** Name: Something concise and clear (which could very well be the name you're currently using). It doesn't need to be the same name that jaws uses - this is easily changed in the above transformations.
*** The value range: That is, values that would be sensible for a commonly used version of setting/preference. It doesn't have to fit the values that jaws uses exactly - the capabilities transformations are quite flexible when it comes to this.
*** Description: This will allow other people to understand what the setting means and when it's used. 
*** Default value: Again, this doesn't have to be the same as jaws, but it likely will, unless jaws has some odd choices on defaults.

I realize that the above is a bit of work, but we need these things (like value ranges and default values) specified for any new common term that is defined. Also, if you're in doubt about whether a setting is sensible to have as a common term or not, I suggest you leave it out for now - it's easy to create a new common term at a later point - and we might have a better sense of which are useful as we integrate more solutions.

kaspermarkus replied:
I know these decisions on what makes a good common term, and value range, etc., are quite difficult, so if you have questions, feel very free to ask for any clarification/guidance necessary. Alternatively, if you want to, we could do a skype call or something to talk through some of them.

kaspermarkus commented on tests/platform/windows/windows-jaws-testSpec.txt:
```diff
@@ -5,6 +5,6 @@ This will run the acceptance tests for the JAWS screenreader (Freedom Scientific
 * JAWS needs to be installed
-* In JAWS, a profile should be created by going to "options"->voices->Voice Adjustment. Make sure that "Eloquence - (factory)" is selected in the profile name. Click "Save As.." button and write "cloud4all" as the name (no quotes). Once this is done, Jaws should be working properly
+* In JAWS, a profile should be created by going to "options"->voices->Voice Adjustment. Make sure that "Eloquence - (factory)" is selected in the profile name. Click "Save As.." button and write "GPII" as the name (no quotes). Once this is done, Jaws should be working properly
 
```
Would you happen to know of any way to do this automatically. Is there some sort of wizard or the like, that would allow us to create a new voice profile automatically via the GPII? Most important is that we would really want to be able to create this profile without requiring any user interaction. So the when the GPII wanted to run JAWS, it would check if there is a GPII voice profile, and if not, it would automaticlly create it.

There might also be alternative ways of setting the voice settings, without requiring to create a new voice profile. Basically, anything that would allow us to edit the Jaws settings without first have to manually do some steps would be extremely useful.

xpheal replied:
Yes, there's a way to do this automatically.

JAWS has two settings file: the main settings file "DEFAULT.jcf" and the voice profile setting "xxx.JCF". We just need to change the Voice "Profiles.ActiveVoiceProfileName = xxx" setting in "DEFAULT.jcf". "xxx" will be the name of the voice profile setting.

Currently, I'm setting the Voice "Profiles.ActiveVoiceProfileName = GPII"and naming the voice profile "GPII.vpf". You will also have to set this setting in "Options.PrimarySynthesizer = eloq" in the voice profile settings.

The problem that I'm facing now is that GPII won't create the settings file if it does not exist. Here's the error I got : {"isError":true,"message":"ENOENT, no such file or directory 'C:\\Users\\user\\AppData\\Roaming\\Freedom Scientific\\JAWS\\17.0\\Settings\\VoiceProfiles\\GPII.VPF'"}
I think this can be solved by adding a few lines of code in the main settings handler code to allow file creation if it does not exist.

There might be other alternative ways. JAWS has an API and I'm looking into it but it requires running a script.

kaspermarkus commented:
Hi @xpheal , It's been almost a year to the day since this pull request was committed (or last updated?) so I think it's about time it got some proper reviewing - with lots of apologies for the complete unresponsiveness from the core team.

I'll be the reviewer, and as you can see, I've already made some comments on the pull request .. Please feel very free to, at any time, ask for clarifcations/help/support/a skype call, or whatever would be helpful to get you through getting this pull request ready for merging. Also, feedback on things that are difficult, need documentation, etc., is always helpful

kaspermarkus commented:
This is out of sync with master, after antranigs big branch merge, so if you could merge it with the latest updates from the GPII master branch, that would be great!

gloob commented:
Thank you @kaspermarkus you are just in time for the celebration of the first birthday of this PR ;)

xpheal commented:
Hi @kaspermarkus, I've updated the github repo with the changes.
Changes made:
1) Merge with latest master branch
2) Solve the language settings problem
3) Sort through the settings in JAWS and implemented only the useful common settings in the solution
4) You still have to create the two settings files in order for the solution to work because JAWS does not create the settings file on installation and GPII can't create it automatically. But, you don't have to go through the trouble to launch JAWS and set up the voice profile manually. It'll be great if we can allow GPII to create the files if they do not exist.

kaspermarkus commented on testData/deviceReporter/installedSolutions.json:
```diff
@@ -1,5 +1,9 @@
 [
     {
+        "id": "com.freedomscientific.jaws"
```
This can be removed again. We made a decision a little while back that the installed Solutions file should only contain things that will be installed by default on the device (i.e. built-in features) - since we risk the system crashes if we attempt to run Jaws without it being installed.. If one wants to do development with jaws, one can always add it to their own installedSolutions file .

On non-development machines, we're using the dynamic device reporter (which uses the "isInstalled" block to decide whether it's installed) - so for these, the installedSolutions file doesn't matter

kaspermarkus commented on tests/platform/windows/windows-jaws-testSpec.js:
```diff
@@ -47,19 +68,64 @@ gpii.tests.windows.jaws = [
+                            "filename": "${{environment}.APPDATA}\\Freedom Scientific\\JAWS\\17.0\\Settings\\enu\\DEFAULT.JCF"
                         }
                     }
                 ]
```
This is awesome! It would be awesome to test some other values for the common terms (ie. to test more cases in your transformations)  - could you perhaps add one more acceptance test (ie. another NP set, and another block with expected results in this file) that contains some different values for the preferences - just to get a bit broader on the testing

kaspermarkus commented:
Great work @xpheal ! I've added a few more comments, and am looking into adding the functionality for automatically creating the files if they dont already exist. Also, I'm talking with antranig about the possiblity of making a new transformation function, that will allow you to take a single value and output to multiple locations - having one of those would clean up your transformations nicely!

Anyway - I'll be doing more checking and testing now, but so far this is looking really good! It's gonna be nice to get into master

kaspermarkus commented:
ok to test

kaspermarkus commented:
Just an update on the status: I've tested most of it and things are looking quite good. There is an issue with one of the acceptance tests, which is failing for me - the one with jaws_application user token. When testing this manually, I see that one problem we got with Jaws is this: https://issues.gpii.net/browse/GPII-1223 . I will talk with @amb26 about how to deal with that issue

kaspermarkus commented:
Updated to latest master and made some changes to work with the latest model transformation framework updates, etc... This has all been pushed here: #481 and has been assigned to Javi for review

kaspermarkus commented:
Closing as all the work from this branch is now part of #481
